Begin integrating WebAuthn with LoginStatus API

pooneh-nb opened the pull request:
#### 39dc84ffd6e37c66dfed5af48c013ee4561bab93
<pre>
Begin integrating WebAuthn with LoginStatus API
<a href="https://bugs.webkit.org/show_bug.cgi?id=277359">https://bugs.webkit.org/show_bug.cgi?id=277359</a>
<a href="https://rdar.apple.com/132824596">rdar://132824596</a>

Reviewed by Charlie Wolfe, Chris Dumez, and Pascoe.

We integrate the WebAuthn with LoginStatus API. So setting a status is now conditional to logging in using WebAuthn in 30s.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
(WebCore::AuthenticatorCoordinator::discoverFromExternalSource):
* Source/WebCore/page/LoginStatus.cpp:
(WebCore::LoginStatus::LoginStatus):
* Source/WebCore/page/LoginStatus.h:
(WebCore::LoginStatus::domain const):
(WebCore::LoginStatus::tokenType const):
(WebCore::LoginStatus::authType const):
(WebCore::LoginStatus::timeToLive const):
(WebCore::LoginStatus::registrableDomain const): Deleted.
(WebCore::LoginStatus::credentialTokenType const): Deleted.
(WebCore::LoginStatus::authenticationType const): Deleted.
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setLastAuthentication):
* Source/WebCore/page/Page.h:
(WebCore::Page::lastAuthentication const):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::setLoginStatus):
(WebKit::WebResourceLoadStatisticsStore::isLoggedIn):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::setLoginStatus):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
(WebKit::NetworkConnectionToWebProcess::isLoginStatusAPIRequiresWebAuthnEnabled const):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setLoginStatus):
* Tools/TestWebKitAPI/Tests/WebCore/LoginStatus.cpp:
(TestWebKitAPI::TEST(LoginStatus, DefaultExpiryWebAuthn)):
(TestWebKitAPI::TEST(LoginStatus, DefaultExpiryPasswordManager)):
(TestWebKitAPI::TEST(LoginStatus, DefaultExpiryUnmanaged)):
(TestWebKitAPI::TEST(LoginStatus, CustomExpiryBelowLong)):
(TestWebKitAPI::TEST(LoginStatus, CustomExpiryBelowShort)):
(TestWebKitAPI::TEST(LoginStatus, RenewedExpiry)):
(TestWebKitAPI::TEST(LoginStatus, ClampedExpiryLong)):
(TestWebKitAPI::TEST(LoginStatus, ClampedExpiryShort)):

Canonical link: <a href="https://commits.webkit.org/283025@main">https://commits.webkit.org/283025@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d3e18c66806235ef7f0a59130989818365fdf331

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64887 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44252 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17491 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68911 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15493 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67004 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52033 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15772 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52140 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10694 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67953 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40918 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56146 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32760 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37589 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13516 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14369 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/57997 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59498 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13852 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70616 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/64128 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8832 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13337 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59471 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8866 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56215 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/59681 "Found 2 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/selections, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/basic-hierarchy (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7287 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/958 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/85896 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9850 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40064 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15138 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41142 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42323 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40884 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pascoej commented on Source/WebCore/page/LoginStatus.h:
```diff
@@ -37,25 +37,29 @@ namespace WebCore {
 
-    WEBCORE_EXPORT static ExceptionOr<UniqueRef<LoginStatus>> create(const RegistrableDomain&, const String& username, CredentialTokenType, AuthenticationType);
-    WEBCORE_EXPORT static ExceptionOr<UniqueRef<LoginStatus>> create(const RegistrableDomain&, const String& username, CredentialTokenType, AuthenticationType, Seconds timeToLive);
+    WEBCORE_EXPORT static ExceptionOr<UniqueRef<LoginStatus>> create(const RegistrableDomain&, const String&, CredentialTokenType, AuthenticationType);
```
We can keep the name for username, since the name adds more information than just the type.

pascoej commented on Source/WebCore/page/LoginStatus.h:
```diff
@@ -37,25 +37,29 @@ namespace WebCore {
-    WEBCORE_EXPORT static ExceptionOr<UniqueRef<LoginStatus>> create(const RegistrableDomain&, const String& username, CredentialTokenType, AuthenticationType);
-    WEBCORE_EXPORT static ExceptionOr<UniqueRef<LoginStatus>> create(const RegistrableDomain&, const String& username, CredentialTokenType, AuthenticationType, Seconds timeToLive);
+    WEBCORE_EXPORT static ExceptionOr<UniqueRef<LoginStatus>> create(const RegistrableDomain&, const String&, CredentialTokenType, AuthenticationType);
+    WEBCORE_EXPORT static ExceptionOr<UniqueRef<LoginStatus>> create(const RegistrableDomain&, const String&, CredentialTokenType, AuthenticationType, Seconds);
```
"

pascoej commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
```diff
@@ -112,11 +115,14 @@ class WebResourceLoadStatisticsStore final : public ThreadSafeRefCountedAndCanMa
 
     ~WebResourceLoadStatisticsStore();
 
+    // WebPage& webPage() const { return m_webPage;}
```
remove

pascoej commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
```diff
@@ -257,12 +263,14 @@ class WebResourceLoadStatisticsStore final : public ThreadSafeRefCountedAndCanMa
 
     RunLoop::Timer m_dailyTasksTimer;
 
+    // WebPage& m_webPage;
```
remove

pascoej commented on Source/WebCore/page/LoginStatus.h:
```diff
@@ -37,25 +37,29 @@ namespace WebCore {
 class LoginStatus {
     WTF_MAKE_ISO_ALLOCATED_EXPORT(LoginStatus, WEBCORE_EXPORT);
 public:
+
```
remove

pascoej reviewed (commented):
Although we could send the WebAuthn signal from the UI process, I don't think there's clear benefit right now.

This looks good other than a few comments.

charliewolfe commented on LayoutTests/http/wpt/webauthn/public-key-credential-get-success-hid.https.html:
The changes in this file should be undone.

charliewolfe commented on LayoutTests/platform/mac/http/tests/login-status/login-status-api-getter-setter-functions-expected.png:
Remove this file.

charliewolfe commented on Source/WebCore/page/Page.cpp:
```diff
@@ -4986,4 +4986,12 @@ void Page::activeNowPlayingSessionUpdateTimerFired()
 
+void Page::setLastAuthentication(LoginStatus::AuthenticationType authType)
+{
+    auto loginStatus = LoginStatus::create(RegistrableDomain(mainFrameURL()), "test"_s, WebCore::LoginStatus::CredentialTokenType::HTTPStateToken , authType, WebCore::LoginStatus::TimeToLiveAuthentication);
```
I don't think we should land this with `"test"` always being the username we pass here. If it's easy to get the username from the caller of this function we should do that. Otherwise we should have a way to pass an empty/null string for now.

charliewolfe commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
Undo the changes in this file.

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -8334,7 +8334,11 @@ void WebPage::requestStorageAccess(RegistrableDomain&& subFrameDomain, Registrab
 void WebPage::setLoginStatus(RegistrableDomain&& domain, IsLoggedIn loggedInStatus, CompletionHandler<void()>&& completionHandler)
 {
-    WebProcess::singleton().ensureNetworkProcessConnection().protectedConnection()->sendWithAsyncReply(Messages::NetworkConnectionToWebProcess::SetLoginStatus(WTFMove(domain), loggedInStatus), WTFMove(completionHandler));
+    RefPtr<WebCore::Page> page = corePage();
```
```suggestion
    RefPtr page = corePage();
```

charliewolfe commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -1190,11 +1190,15 @@ void NetworkConnectionToWebProcess::requestStorageAccess(RegistrableDomain&& sub
 {
+    if (isLoginStatusAPIRequiresAuthnEnabled() && !lastAuthentication) {
+        return completionHandler();
+    }
```
```suggestion
    if (isLoginStatusAPIRequiresAuthnEnabled() && !lastAuthentication)
        return completionHandler();
```

charliewolfe commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
```diff
@@ -50,6 +50,8 @@
 #include <wtf/text/WTFString.h>
 
 namespace WebCore {
+
```
Remove extra new line.

charliewolfe commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
```diff
@@ -62,6 +64,7 @@ namespace WebKit {
 class NetworkSession;
 class ResourceLoadStatisticsStore;
 class WebFrameProxy;
+class WebPage;
```
This forward declaration is not needed.

charliewolfe commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -4160,6 +4160,21 @@ LoginStatusAPIEnabled:
+  type: bool
+  status: testable
+  category: dom
+  humanReadableName: "Require Login Status API with WebAuthn"
```
```suggestion
  humanReadableName: "Require WebAuthn with the Login Status API"
```

charliewolfe commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -4160,6 +4160,21 @@ LoginStatusAPIEnabled:
     WebCore:
       default: false
 
+LoginStatusAPIRequiresAuthnEnabled:
```
```suggestion
LoginStatusAPIRequiresWebAuthnEnabled:
```

charliewolfe commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -4160,6 +4160,21 @@ LoginStatusAPIEnabled:
+  status: testable
+  category: dom
+  humanReadableName: "Require Login Status API with WebAuthn"
+  humanReadableDescription: "Enable the proposed Login Status API which requires WebAuthn"
```
This preference doesn't enable the Login Status API on its own, we should change the description to reflect that.

charliewolfe commented on Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
```diff
@@ -340,13 +342,15 @@ void AuthenticatorCoordinator::discoverFromExternalSource(const Document& docume
 
         if (auto response = AuthenticatorResponse::tryCreate(WTFMove(data), attachment)) {
+            if (RefPtr document = weakDocument.get())
+                document->page()->setLastAuthentication(LoginStatus::AuthenticationType::WebAuthn);
```
`page()` needs a null check here

charliewolfe commented on Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
```diff
@@ -340,13 +342,19 @@ void AuthenticatorCoordinator::discoverFromExternalSource(const Document& docume
+                if (!page)
+                    return;
+                page->setLastAuthentication(LoginStatus::AuthenticationType::WebAuthn);
+            }
```
You can just capture page instead of document in this lambda: `weakPage = WeakPtr { document.page() }`.

Then this could be changed to:
```suggestion
            if (RefPtr page = weakPage.get())
                page->setLastAuthentication(LoginStatus::AuthenticationType::WebAuthn);
```

charliewolfe commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -4189,6 +4189,21 @@ LoginStatusAPIEnabled:
+  status: internal
+  category: dom
+  humanReadableName: "Require WebAuthn with the Login Status API"
+  humanReadableDescription: "Enable the proposed Login Status API which requires WebAuthn"
```
This preference doesn't actually enable the Login Status API, it just makes it so WebAuthn is required for it to work. So maybe this description should be something like:
```suggestion
  humanReadableDescription: "Require a recent WebAuthn authentication to set login status"
```

charliewolfe commented on Source/WebCore/page/LoginStatus.h:
```diff
@@ -38,24 +38,27 @@ class LoginStatus {
 
     WEBCORE_EXPORT static ExceptionOr<UniqueRef<LoginStatus>> create(const RegistrableDomain&, const String& username, CredentialTokenType, AuthenticationType);
-    WEBCORE_EXPORT static ExceptionOr<UniqueRef<LoginStatus>> create(const RegistrableDomain&, const String& username, CredentialTokenType, AuthenticationType, Seconds timeToLive);
+    WEBCORE_EXPORT static ExceptionOr<UniqueRef<LoginStatus>> create(const RegistrableDomain&, const String& username, CredentialTokenType, AuthenticationType, Seconds);
```
This change should be undone. The variable name should be included in the function declaration unless it's obvious from the type name.

charliewolfe commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
Undo the changes in this file

pascoej commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -1191,11 +1191,14 @@ void NetworkConnectionToWebProcess::requestStorageAccess(RegistrableDomain&& sub
-void NetworkConnectionToWebProcess::setLoginStatus(RegistrableDomain&& domain, IsLoggedIn loggedInStatus, CompletionHandler<void()>&& completionHandler)
+void NetworkConnectionToWebProcess::setLoginStatus(RegistrableDomain&& domain, IsLoggedIn loggedInStatus, std::optional<WebCore::LoginStatus> lastAuthentication, CompletionHandler<void()>&& completionHandler)
 {
+    if (isLoginStatusAPIRequiresAuthnEnabled() && !lastAuthentication)
```
I don't think we should reject here. We can still try to set the login status, but avoid setting the internal "userMediated" flag for this authentication (in other words set authenticationType=unmanaged. 


We might want to reject here if we implement the "userMediated" option to pass to setStatus. If that flag was passed by the RP and contains true, we would reject any requests without a browser mediated ceremony proceeding it (what you are doing here currently.)

pooneh-nb replied:
But we don't have the "userMediated" option yet. Should I consider to implement that option as well? or just changing the authentication type is enough?

pascoej replied:
I think just changing the authentication type for now is enough. We shouldn't error out of the request.

charliewolfe commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -1191,11 +1191,18 @@ void NetworkConnectionToWebProcess::requestStorageAccess(RegistrableDomain&& sub
+    if (isLoginStatusAPIRequiresAuthnEnabled() && !lastAuthentication) {
+    auto umanagedAuthentication = LoginStatus::create(domain, emptyString(), WebCore::LoginStatus::CredentialTokenType::HTTPStateToken , WebCore::LoginStatus::AuthenticationType::Unmanaged, WebCore::LoginStatus::TimeToLiveAuthentication);
+    if (umanagedAuthentication.hasException())
+        return;
```
The completion handler should be called here, since we are returning.
```suggestion
        return completionHandler();
```

charliewolfe commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -1191,11 +1191,18 @@ void NetworkConnectionToWebProcess::requestStorageAccess(RegistrableDomain&& sub
+    if (umanagedAuthentication.hasException())
+        return;
+    lastAuthentication = umanagedAuthentication.returnValue();
+    }
```
The content inside the if block should be indented

charliewolfe commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -1191,11 +1191,18 @@ void NetworkConnectionToWebProcess::requestStorageAccess(RegistrableDomain&& sub
+void NetworkConnectionToWebProcess::setLoginStatus(RegistrableDomain&& domain, IsLoggedIn loggedInStatus, std::optional<WebCore::LoginStatus> lastAuthentication, CompletionHandler<void()>&& completionHandler)
 {
+    if (isLoginStatusAPIRequiresAuthnEnabled() && !lastAuthentication) {
+    auto umanagedAuthentication = LoginStatus::create(domain, emptyString(), WebCore::LoginStatus::CredentialTokenType::HTTPStateToken , WebCore::LoginStatus::AuthenticationType::Unmanaged, WebCore::LoginStatus::TimeToLiveAuthentication);
```
```suggestion
    auto umanagedAuthentication = LoginStatus::create(domain, emptyString(), LoginStatus::CredentialTokenType::HTTPStateToken, LoginStatus::AuthenticationType::Unmanaged, LoginStatus::TimeToLiveAuthentication);
```

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -8355,7 +8355,11 @@ void WebPage::requestStorageAccess(RegistrableDomain&& subFrameDomain, Registrab
-    WebProcess::singleton().ensureNetworkProcessConnection().protectedConnection()->sendWithAsyncReply(Messages::NetworkConnectionToWebProcess::SetLoginStatus(WTFMove(domain), loggedInStatus), WTFMove(completionHandler));
+    RefPtr page = corePage();
+    if (!page)
+        return;
```
completion handler should be called here too:
```suggestion
        return completionHandler();
```

charliewolfe commented on Source/WebCore/page/Page.cpp:
```diff
@@ -5030,4 +5030,12 @@ void Page::activeNowPlayingSessionUpdateTimerFired()
 
+void Page::setLastAuthentication(LoginStatus::AuthenticationType authType)
+{
+    auto loginStatus = LoginStatus::create(RegistrableDomain(mainFrameURL()), emptyString(), WebCore::LoginStatus::CredentialTokenType::HTTPStateToken , authType, WebCore::LoginStatus::TimeToLiveAuthentication);
```
```suggestion
    auto loginStatus = LoginStatus::create(RegistrableDomain(mainFrameURL()), emptyString(), LoginStatus::CredentialTokenType::HTTPStateToken, authType, LoginStatus::TimeToLiveAuthentication);
```

charliewolfe commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
```diff
@@ -111,4 +111,4 @@ messages -> NetworkProcessProxy LegacyReceiver {
 }
 
-}
+}
```
Undo this change

charliewolfe commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
```diff
@@ -365,8 +366,9 @@ class NetworkConnectionToWebProcess
-    void setLoginStatus(RegistrableDomain&&, WebCore::IsLoggedIn, CompletionHandler<void()>&&);
+    void setLoginStatus(RegistrableDomain&&, WebCore::IsLoggedIn, std::optional<WebCore::LoginStatus>, CompletionHandler<void()>&&);
     void isLoggedIn(RegistrableDomain&&, CompletionHandler<void(bool)>&&);
+    bool isLoginStatusAPIRequiresAuthnEnabled() const { return m_sharedPreferencesForWebProcess.loginStatusAPIRequiresWebAuthnEnabled; }
```
We should update this function name to match the name of the setting:
```suggestion
    bool isLoginStatusAPIRequiresWebAuthnEnabled() const { return m_sharedPreferencesForWebProcess.loginStatusAPIRequiresWebAuthnEnabled; }
```

charliewolfe commented on Source/WebCore/page/Page.h:
```diff
@@ -33,6 +33,7 @@
 #include "LengthBox.h"
 #include "LoadSchedulingMode.h"
 #include "LocalFrame.h"
+#include "LoginStatus.h"
```
Can you forward declare LoginStatus and move this include to the cpp file? We try to avoid adding includes in headers to reduce build times.

charliewolfe replied:
Actually, I'm wrong about this - you can't forward declare here. Sorry about that.

cdumez commented on Source/WebCore/page/Page.cpp:
```diff
@@ -5030,4 +5031,12 @@ void Page::activeNowPlayingSessionUpdateTimerFired()
+    auto loginStatus = LoginStatus::create(RegistrableDomain(mainFrameURL()), emptyString(), LoginStatus::CredentialTokenType::HTTPStateToken, authType, LoginStatus::TimeToLiveAuthentication);
+    if (loginStatus.hasException())
+        return;
+    m_lastAuthentication = loginStatus.returnValue();
```
`loginStatus.releaseReturnValue()` to avoid copying.

cdumez commented on Source/WebCore/page/Page.h:
```diff
@@ -1200,6 +1201,9 @@ class Page : public RefCounted<Page>, public Supplementable<Page>, public CanMak
 #endif
 
+    void setLastAuthentication(LoginStatus::AuthenticationType);
+    std::optional<LoginStatus> lastAuthentication() const { return m_lastAuthentication; }
```
should return a `const std::optional<LoginStatus>&` to not force a copy.

cdumez commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -444,13 +445,17 @@ void WebResourceLoadStatisticsStore::requestStorageAccess(RegistrableDomain&& su
+        if (loginStatusToSet)
+            loginStatusToSet->setTimeToLive(WebCore::LoginStatus::TimeToLiveLong);
+        auto pair = std::make_pair(loggedInStatus, loginStatusToSet);
+        m_loginStatus.set(domain, pair);
```
`WTFMove(pair)`

cdumez commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -1191,11 +1191,18 @@ void NetworkConnectionToWebProcess::requestStorageAccess(RegistrableDomain&& sub
+        auto umanagedAuthentication = LoginStatus::create(domain, emptyString(), WebCore::LoginStatus::CredentialTokenType::HTTPStateToken, WebCore::LoginStatus::AuthenticationType::Unmanaged, WebCore::LoginStatus::TimeToLiveAuthentication);
+        if (umanagedAuthentication.hasException())
+            return completionHandler();
+        lastAuthentication = umanagedAuthentication.returnValue();
```
`releaseReturnValue()`

cdumez commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -1191,11 +1191,18 @@ void NetworkConnectionToWebProcess::requestStorageAccess(RegistrableDomain&& sub
 }
 
-void NetworkConnectionToWebProcess::setLoginStatus(RegistrableDomain&& domain, IsLoggedIn loggedInStatus, CompletionHandler<void()>&& completionHandler)
+void NetworkConnectionToWebProcess::setLoginStatus(RegistrableDomain&& domain, IsLoggedIn loggedInStatus, std::optional<WebCore::LoginStatus> lastAuthentication, CompletionHandler<void()>&& completionHandler)
```
`std::optional<WebCore::LoginStatus>&&`

cdumez commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -1191,11 +1191,18 @@ void NetworkConnectionToWebProcess::requestStorageAccess(RegistrableDomain&& sub
     if (auto* networkSession = this->networkSession()) {
         if (auto* resourceLoadStatistics = networkSession->resourceLoadStatistics()) {
-            resourceLoadStatistics->setLoginStatus(WTFMove(domain), loggedInStatus, WTFMove(completionHandler));
+            resourceLoadStatistics->setLoginStatus(WTFMove(domain), loggedInStatus, lastAuthentication, WTFMove(completionHandler));
```
`WTFMove(lastAuthentication)`

cdumez commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
```diff
@@ -144,7 +146,7 @@ class WebResourceLoadStatisticsStore final : public ThreadSafeRefCountedAndCanMa
     bool hasStorageAccessForFrame(const SubFrameDomain&, const TopFrameDomain&, WebCore::FrameIdentifier, WebCore::PageIdentifier);
     void requestStorageAccess(SubFrameDomain&&, TopFrameDomain&&, WebCore::FrameIdentifier, WebCore::PageIdentifier, WebPageProxyIdentifier, StorageAccessScope,  CompletionHandler<void(RequestStorageAccessResult)>&&);
-    void setLoginStatus(RegistrableDomain&&, IsLoggedIn, CompletionHandler<void()>&&);
+    void setLoginStatus(RegistrableDomain&&, IsLoggedIn, std::optional<LoginStatus>, CompletionHandler<void()>&&);
```
`std::optional<LoginStatus>&&`

cdumez commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -444,13 +445,17 @@ void WebResourceLoadStatisticsStore::requestStorageAccess(RegistrableDomain&& su
 }
 
-void WebResourceLoadStatisticsStore::setLoginStatus(RegistrableDomain&& domain, IsLoggedIn loggedInStatus, CompletionHandler<void()>&& completionHandler)
+void WebResourceLoadStatisticsStore::setLoginStatus(RegistrableDomain&& domain, IsLoggedIn loggedInStatus, std::optional<LoginStatus> lastAuthentication, CompletionHandler<void()>&& completionHandler)
```
`std::optional<LoginStatus>&&`

cdumez commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -444,13 +445,17 @@ void WebResourceLoadStatisticsStore::requestStorageAccess(RegistrableDomain&& su
-        m_loginStatus.set(domain, loggedInStatus);
-    else
+    if (loggedInStatus == IsLoggedIn::LoggedIn) {
+        auto loginStatusToSet = lastAuthentication && lastAuthentication->hasExpired() ? std::nullopt : std::optional(lastAuthentication);
```
`WTFMove(lastAuthentication)`

cdumez commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
```diff
@@ -365,8 +366,9 @@ class NetworkConnectionToWebProcess
     void requestStorageAccessUnderOpener(WebCore::RegistrableDomain&& domainInNeedOfStorageAccess, WebCore::PageIdentifier openerPageID, WebCore::RegistrableDomain&& openerDomain);
 
-    void setLoginStatus(RegistrableDomain&&, WebCore::IsLoggedIn, CompletionHandler<void()>&&);
+    void setLoginStatus(RegistrableDomain&&, WebCore::IsLoggedIn, std::optional<WebCore::LoginStatus>, CompletionHandler<void()>&&);
```
`std::optional<WebCore::LoginStatus>&&`

Commit:
Begin integrating WebAuthn with LoginStatus API
https://bugs.webkit.org/show_bug.cgi?id=277359
rdar://132824596

Reviewed by Charlie Wolfe, Chris Dumez, and Pascoe.

We integrate the WebAuthn with LoginStatus API. So setting a status is now conditional to logging in using WebAuthn in 30s.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
(WebCore::AuthenticatorCoordinator::discoverFromExternalSource):
* Source/WebCore/page/LoginStatus.cpp:
(WebCore::LoginStatus::LoginStatus):
* Source/WebCore/page/LoginStatus.h:
(WebCore::LoginStatus::domain const):
(WebCore::LoginStatus::tokenType const):
(WebCore::LoginStatus::authType const):
(WebCore::LoginStatus::timeToLive const):
(WebCore::LoginStatus::registrableDomain const): Deleted.
(WebCore::LoginStatus::credentialTokenType const): Deleted.
(WebCore::LoginStatus::authenticationType const): Deleted.
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setLastAuthentication):
* Source/WebCore/page/Page.h:
(WebCore::Page::lastAuthentication const):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::setLoginStatus):
(WebKit::WebResourceLoadStatisticsStore::isLoggedIn):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::setLoginStatus):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
(WebKit::NetworkConnectionToWebProcess::isLoginStatusAPIRequiresWebAuthnEnabled const):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setLoginStatus):
* Tools/TestWebKitAPI/Tests/WebCore/LoginStatus.cpp:
(TestWebKitAPI::TEST(LoginStatus, DefaultExpiryWebAuthn)):
(TestWebKitAPI::TEST(LoginStatus, DefaultExpiryPasswordManager)):
(TestWebKitAPI::TEST(LoginStatus, DefaultExpiryUnmanaged)):
(TestWebKitAPI::TEST(LoginStatus, CustomExpiryBelowLong)):
(TestWebKitAPI::TEST(LoginStatus, CustomExpiryBelowShort)):
(TestWebKitAPI::TEST(LoginStatus, RenewedExpiry)):
(TestWebKitAPI::TEST(LoginStatus, ClampedExpiryLong)):
(TestWebKitAPI::TEST(LoginStatus, ClampedExpiryShort)):

Canonical link: https://commits.webkit.org/283025@main
